### ISSUE TITLE
[HLSL] Add support for `SPV_NV_mesh_shader`

### ISSUE THREAD

**Issue by Try** [KhronosGroup/SPIRV-Cross#1974]
DirectX12 now has similar(more rich actually) functionality out of the box, so it would be nice to translate `mesh` and `task` shaders.
Links:
https://github.com/KhronosGroup/GLSL/blob/master/extensions/nv/GLSL_NV_mesh_shader.txt
https://microsoft.github.io/DirectX-Specs/d3d/MeshShader.html#checkfeaturesupport

Core functionality is almost identical, but there are corner-cases with `points` and others.

#### General
 1. `local_size_*` and `[numthreads]`
   Vulkan:  `local_size_x<=32; local_size_y=1; local_size_=1`
   DirectX12: `numthreads X*Y*Z <= 128` 
2. Intrisics:
   `gl_WorkGroupSize`, `gl_WorkGroupID`, and others - same as compute shader. 

#### Mesh
 1. Output data size:
   Vulkan: compile time known by `layout(triangles, max_vertices = *, max_primitives = *) out;` 
   DirectX12:  determinate by array size of `out` data.
 2. Output topology:
   Vulkan: compile time known by `layout(triangles) out`. Possible values are: `points`, `lines`, `triangles`
   DirectX12: compile time known by: `[OutputTopology("triangle")]`. Possible values are: `"lines"`, `"triangles"`

#### Task
Task is mainly compute shader, except special builtin outputs:
1. `gl_TaskCountNV` -> doesn't maps directly, should be passed to `DispatchMesh`
2. `taskNV` -> also parameter to `DispatchMesh`. Mesh stage receives it as `in payload`.
NOTE: tested locally task translation, since it's quite simple: apparently there is no version of `DispatchMesh` in HLSL that doesn't takes `payload_t` input and `payload_t` cannot be empty struct. This brings extra complexity to translation.

#### Builtin variables
1. `gl_WorkGroupSize` ->  N/A, but it's not an issue
4. `gl_WorkGroupID` -> `SV_GroupID`
5. `gl_LocalInvocationID` -> `SV_GroupThreadID`
6. `gl_GlobalInvocationID` -> `SV_DispatchThreadID`
7. `gl_LocalInvocationIndex` -> `SV_GroupIndex`
8. `gl_TaskCountNV` -> `DispatchMesh(count,1,1,usrTaskData)` at tail of `amplification` stage
9. `gl_PrimitiveCountNV` -> [`SetMeshOutputCounts(vertCount, gl_PrimitiveCountNV)`](https://microsoft.github.io/DirectX-Specs/d3d/MeshShader.html#setmeshoutputcounts) before writing outputs, in first active thread.
10. `gl_PrimitiveIndicesNV` -> `out indices uint3 tris[max_primitives/3]` - part of main function signature. Unlike Vulkan write-only.
11. `gl_Position` -> `out vertices MyOutputVertex verts[max_vertices]`; `SV_Position`, as member of output per-vertex data (`MyOutputVertex`)
12. `gl_PositionPerViewNV` 
13. `gl_PointSize` - point aren't supported in DX
14. `gl_ClipDistance`  - not mention in spec, maybe SV_ClipDistance still works 
15. `gl_ClipDistancePerViewNV`
16. `gl_CullDistance` - same as `gl_ClipDistance`
17. `gl_CullDistancePerViewNV`
18. `gl_PrimitiveID` -> `SV_PrimitiveID` (similar to glsl mesh stage should generate primitive id manually)
19. `gl_Layer`
20. `gl_LayerPerViewNV`
21. `gl_ViewportIndex` -> `SV_ViewportArrayIndex`
22. `gl_ViewportMask`
23. `gl_ViewportMaskPerViewNV`
24. `gl_MeshViewCountNV`
25. `gl_MeshViewIndicesNV`
26. `gl_DrawID`
27. `writePackedPrimitiveIndices4x8NV()` - N/A, can be emulated

TODO:  check layered rendering and multiview.

Meshlet outputs:
`out indices uint3 tris[max_primitives/3]` - local index buffer. GLSL: `gl_PrimitiveIndicesNV`
`out vertices MyOutputVertex verts[max_vertices]` - local vertex buffer. GLSL: `gl_MeshVerticesNV`
`out primitives PrimitiveAttributes sharedPrimitives[max_primitives]` - per-primitive data

**Comment by Try** [comment]
EXT is out:
https://www.khronos.org/blog/mesh-shading-for-vulkan

and conveniently they added a mapping table to the post :)

**Comment by Try** [comment]
Been experimenting with `GL_EXT_mesh_shader` on weekend. Few notes/problems when it comes to HLSL translation:

1. HLSL doesn't allow for multiple `out vertices` or `out primitives` declarations
That will cause issues with user defined outputs
2. Currently spirv-cross performs `SV_CullDistance[N]` unrolling. This not going to work as good with arrays.

Question: is it OK to put all user defined vertex/primitive outputs into shared memory and then copy them in shader epilogue?

**Comment by HansKristian-Work** [comment]
> Question: is it OK to put all user defined vertex/primitive outputs into shared memory and then copy them in shader epilogue?

No. A compiler cannot optimize this away.

> HLSL doesn't allow for multiple out vertices or out primitives declarations

If leaf functions access these variables, that might be a problem, I'm not sure if we can pass down the out blocks by reference, normal HLSL semantics imply an inout copy, which breaks.

**Comment by Try** [comment]
> To make this work, we'd have to collect all per-vertex and per-primitives into one block.

This doesn't look like possible solution. If GLSL has something simple, like:
```
layout(location = 0) out vec4 vOut[];
void foo(out vec4 x[24])
{
    x[gl_LocalInvocationIndex] = vec4(vec3(gl_GlobalInvocationID), 2.0);
}
...
foo(vOut); // how to translate it, if vOut is merged with gl_MeshPerVertexEXT and others?
```

> A compiler cannot optimize 

This is not the easiest case for optimizer, but I think worth a shot to see what DXIL actually does. Not to mention NV version will definitely require shared memory based solution, as it allows read/write and generally way more stuff

**Comment by HansKristian-Work** [comment]
> This doesn't look like possible solution. If GLSL has something simple, like:

The problem is:

```
layout(location = 0) out vec4 A[];
layout(location = 1) out vec4 B[];

void inner_function()
{
    A[gl_LocalInvocationIndex] = vec4(1.0);
    B[gl_LocalInvocationIndex] = vec4(2.0);
}

void main()
{
   inner_function();
}
```

I'm not entirely use how that would look in HLSL, you'd need something like:

```
struct VOut
{
	float4 pos : SV_Position;
	float4 b : B;
	float4 c : C;
};

groupshared float foo[64];

struct Payload
{
	float p;
};

void in_inner_leaf1(inout VOut vout[24], uint tid)
{
	vout[tid].b = 1.0.xxxx;
}

void in_inner_leaf2(inout VOut vout[24], uint tid)
{
	vout[tid].c = 3.0.xxxx;
}

void in_leaf(inout VOut vout[24], uint tid)
{
	in_inner_leaf1(vout, tid);
	vout[tid].pos = 2.0.xxxx;
	in_inner_leaf2(vout, tid);
}

[numthreads(2, 3, 4)]
[outputtopology("triangle")]
void main(uint tid : SV_GroupIndex,
		in payload Payload p,
		out vertices VOut vout[24],
		out indices uint3 ind[8])
{
	foo[tid] = float(tid);
	GroupMemoryBarrierWithGroupSync();
	SetMeshOutputCounts(24, 8);
	in_leaf(vout, tid);
	if (tid < 8)
		ind[tid] = 3 * tid + uint3(0, 1, 2);
}
```

This seems to compile into something sensible though, so maybe it is possible to rely on inout.

```
@"\01?foo@@3PAMA" = external addrspace(3) global [64 x float], align 4

define void @main() {
  %1 = call i32 @dx.op.flattenedThreadIdInGroup.i32(i32 96)  ; FlattenedThreadIdInGroup()
  %2 = uitofp i32 %1 to float
  %3 = getelementptr [64 x float], [64 x float] addrspace(3)* @"\01?foo@@3PAMA", i32 0, i32 %1
  store float %2, float addrspace(3)* %3, align 4, !tbaa !16
  call void @dx.op.barrier(i32 80, i32 9)  ; Barrier(barrierMode)
  call void @dx.op.setMeshOutputCounts(i32 168, i32 24, i32 8)  ; SetMeshOutputCounts(numVertices,numPrimitives)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 1, i32 0, i8 0, float 1.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 1, i32 0, i8 1, float 1.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 1, i32 0, i8 2, float 1.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 1, i32 0, i8 3, float 1.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 0, i32 0, i8 0, float 2.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 0, i32 0, i8 1, float 2.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 0, i32 0, i8 2, float 2.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 0, i32 0, i8 3, float 2.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 2, i32 0, i8 0, float 3.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 2, i32 0, i8 1, float 3.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 2, i32 0, i8 2, float 3.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  call void @dx.op.storeVertexOutput.f32(i32 171, i32 2, i32 0, i8 3, float 3.000000e+00, i32 %1)  ; StoreVertexOutput(outputSigId,rowIndex,colIndex,value,vertexIndex)
  %4 = icmp ult i32 %1, 8
  br i1 %4, label %5, label %9

; <label>:5                                       ; preds = %0
  %6 = mul i32 %1, 3
  %7 = add i32 %6, 1
  %8 = add i32 %6, 2
  call void @dx.op.emitIndices(i32 169, i32 %1, i32 %6, i32 %7, i32 %8)  ; EmitIndices(PrimitiveIndex,VertexIndex0,VertexIndex1,VertexIndex2)
  br label %9

; <label>:9                                       ; preds = %5, %0
  ret void
}
```

**Comment by Try** [comment]
> This seems to compile into something sensible though, so maybe it is possible to rely on inout.

Should be fine to use `out` in HLSL:
```
void in_inner_leaf1(out VOut vout[24], uint tid)
{
	vout[tid].b = 1.0.xxxx;
}

```

**Comment by HansKristian-Work** [comment]
> Should be fine to use out in HLSL:

Technically it does not preserve the inputs, but DXC seems to emit the same code either way. inout is the closest we get to actual reference semantics, so that's what we should use.

**Comment by Try** [comment]
Fine, but in anycase back to https://github.com/KhronosGroup/SPIRV-Cross/issues/1974#issuecomment-1274963627,
Still see no way to merge user defined outputs into one block.

**Comment by HansKristian-Work** [comment]
Doesn't the HLSL backend already do that?

E.g.:

```
#version 450

layout(location = 0) out vec4 A;
layout(location = 1) out vec4 B;

void main()
{
	gl_Position = vec4(1.0);
	A = vec4(2.0);
	B = vec4(3.0);
}
```
looks like this in HLSL output:

```
static float4 gl_Position;
static float4 A;
static float4 B;

struct SPIRV_Cross_Output <-- the merged block
{
    float4 A : TEXCOORD0;
    float4 B : TEXCOORD1;
    float4 gl_Position : SV_Position;
};

void vert_main()
{
    gl_Position = 1.0f.xxxx;
    A = 2.0f.xxxx;
    B = 3.0f.xxxx;
}

SPIRV_Cross_Output main()
{
    vert_main();
    SPIRV_Cross_Output stage_output;
    stage_output.gl_Position = gl_Position;
    stage_output.A = A;
    stage_output.B = B;
    return stage_output;
}
```

The new parts would be collecting per-vertex and per-primitive into separate structs. Access chain code would need to be rewritten as well similar to how we currently emit tessellation code in GLSL backend.

**Comment by Try** [comment]
> Access chain code would need to be rewritten as well similar to how we currently emit tessellation code in GLSL backend.

This is an issue. Like in my code-sample.
GLSL may have function, like
`void foo(out data[max_vertex]);`
This function is valid for both `varying` and regular variables, but access-chain should be rewritten only for mesh-version.
Emitting 2 versions of such function, for `foo<mesh>` and `foo<regular>` doesn't seem to be good approach.

**Comment by HansKristian-Work** [comment]
> GLSL may have function, like
void foo(out data[max_vertex]);

No, it is not valid to pass these things as direct parameters in SPIR-V or GLSL.

**Comment by Try** [comment]
Progressing locally on mesh shader. Some feedback/questions:
> If leaf functions access these variables, 

So far work well, with little pre-pass, inspired by how metal backend works with global variables. Generated code, looks like:
```
// gl_MeshPerVertexEXT is merge of all per-vertex outputs
void foo(out gl_MeshPerVertexEXT gl_MeshVerticesEXT[24])
{
    gl_MeshVerticesEXT[gl_LocalInvocationIndex].outputs.a = 5.0f.xxxx;
    gl_MeshVerticesEXT[gl_LocalInvocationIndex].outputs.b = 6.0f.xxxx;
}
```

> `gl_ClipDistance`, `gl_CullDistance`
Do array declaration for now, like `float4 gl_ClipDistance[1] : SV_ClipDistance;`. Output linker signature look compatible to non-arrayed declaration, but need to test.

> layout(location = 2) out BlockOut

Block outputs are tricky, since HLSL backend flattens structures in outputs, like:
```
struct gl_MeshPerVertexEXT
{
    float4 vOut : TEXCOORD0;
    float4 BlockOut_a : TEXCOORD2;
    float4 BlockOut_b : TEXCOORD3;
    float4 gl_Position : SV_Position;
    float4 gl_ClipDistance[1] : SV_ClipDistance;
    float4 gl_CullDistance[1] : SV_CullDistance;
};
```

But I don't remember why exactly spir-v cross does flattening. Probably allocating multiple output registers for struct is good enough.
Small note outside of HLSL: per-vertex output may become a bigger issue in metal3, because of their api: `void mesh<...>::set_vertex(uint I, V v)`

**Comment by Try** [comment]
Current WIP: c22f7ecf4bd149255b5d08178016a13b09654dc5

Generated HLSL: https://shader-playground.timjones.io/a338900ab871b5d80da47ab308dc72cd
(NOTE: gl_ClipDistance is still todo)

**Comment by Try** [comment]
One particular feature was missed in mesh shader translation: `flip_vert_y`
Implement it same way as for vertex shader would be problematic, as `gl_Position` is write-only now.

Current idea is to detect access chain into `gl_Position` and apply flip on the spot. Something like:
```
// gl_MeshVerticesEXT[gl_LocalInvocationIndex].gl_Position = foo();
vec4 tmp = foo();
tmp.y = -tmp.y;
gl_MeshVerticesEXT[gl_LocalInvocationIndex].gl_Position = tmp;
```
Not sure how difficult that would be for partial writes. Something like `.gl_Position.xzw = ...`

**Comment by Try** [comment]
Find an issue today, when testing cross-compiller in my engine:
`SetMeshOutputsEXT` and `SetMeshOutputCounts` do not have same restrictions. 
In [GLSL](https://github.com/KhronosGroup/GLSL/blob/master/extensions/ext/GLSL_EXT_mesh_shader.txt):
```
...
The arguments are taken from the first invocation in each workgroup.
Any invocation must call this function no more than once and under
uniform control flow, otherwise behavior is undefined. There must not
be any control flow path to an output write that is not preceded by a
call to this function, otherwise behavior is undefined.
```
In [HLSL](https://microsoft.github.io/DirectX-Specs/d3d/MeshShader.html#setmeshoutputcounts):
```
1. This function can only be called once per shader.
...
5. This call must dominate all writes to shared output arrays. 
In other words, there must not be any execution path that even appears
to reach any writes to any output array without first having executed this call.
```

In practice this shader:
```c++
if (uniform_cond) {
  SetMeshOutputCounts(0,0); // leaving without it causes UB on Vulkan side 
  return;
  }
SetMeshOutputCounts(...);
... write-out positions/varyings here ...
```
Is valid from GLSL point of view, but in DXIL:
```
error: SetMeshOUtputCounts cannot be called multiple times.
note: at 'call void @dx.op.setMeshOutputCounts(i32 168, i32 64, i32 %240)' in block '#22' of function 'main'.
Validation failed.
```

Also I'm not 100% sure about HLSL

**Comment by Try** [comment]
Linking DXC ticket about empty task-payload support: https://github.com/microsoft/DirectXShaderCompiler/issues/5269
